contracts/finance/andromeda-validator-staking/src/contract.rs

```diff
@@ -104,6 +105,50 @@ fn execute_stake(ctx: ExecuteContext, validator: Option<Addr>) -> Result<Respons
     Ok(res)
 }
 
+fn execute_unstake(
+    ctx: ExecuteContext,
+    validator: Option<Addr>,
+) -> Result<Response, ContractError> {
+    let ExecuteContext {
+        deps, info, env, ..
+    } = ctx;
+
+    let delegator = env.contract.address;
+    // Ensure sender is the contract owner
+    ensure!(
+        ADOContract::default().is_contract_owner(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {}
+    );
+
+    let default_validator = DEFAULT_VALIDATOR.load(deps.storage)?;
+    let validator = validator.unwrap_or(default_validator);
+
+    // Check if the validator is valid before unstaking
+    is_validator(&deps, &validator)?;
+
+    let Some(res) = deps.querier.query_delegation(delegator.to_string(), validator.to_string())? else {
+        return Err(ContractError::InvalidValidatorOperation { operation: "Unstake".to_string(), validator: validator.to_string() });
+    };
+
+    if res.amount.amount.u128() == 0 {
+        return Err(ContractError::InvalidValidatorOperation {
+            operation: "Unstake".to_string(),
+            validator: validator.to_string(),
+        });
+    }
+
+    let res = Response::new()
```

Comment:
The issue here is that delegation has a variable unbonding period.

How do we know how much the user can claim? Or when they can claim it?

Reply:
The unbonding period is the amount of time that a delegator waits before accessing their undelegated tokens. Thus we need another issue for withdraw method where we verify if delegators can claim the unstaked tokens.

Reply:
#295 has a discussion piece for this already